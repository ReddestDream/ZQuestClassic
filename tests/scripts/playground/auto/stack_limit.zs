#include "auto/test_runner.zs"

void call_recursive(int i)
{
	if (i > 0)
		call_recursive(i - 1);
	Trace(i);
}

generic script stack_limit
{
	void run()
	{
		Test::Init();

		Waitframe();

		// OK.
		Trace("call_recursive: 100");
		call_recursive(100);

		// This should overflow.
		Trace("call_recursive: 10000");
		call_recursive(10000);

		// This won't be hit (but `Test::End()` will still be called by TestRunner).
		Test::Fail("expected overflow");
	}
}
