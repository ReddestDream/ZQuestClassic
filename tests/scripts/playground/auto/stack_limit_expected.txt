stderr:

Compiling 'stack_limit.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Checking code paths
Pass 6: Generating object code
Pass 7: Assembling
Success!
Compile finished with exit code '0' (success)

stdout:

{
  "success": true,
  "diagnostics": [],
  "metadata": {
    "currentFileSymbols": 3,
    "symbols": 9,
    "identifiers": 16,
    "elided": true
  }
}

zasm:

QUIT; void run() Body
PUSHR REFDMAPDATA; void run() Body Start
PUSHARGSV 0 0.0003
SETR D4 SP2
ALLOCATEMEMV D2 100 0
STORE_OBJECT D2 0.0002
LOAD D2 0.0002; InlineFunc[void Game::GetSaveName(char32[])] Params
GETSAVENAME D2; InlineFunc[void Game::GetSaveName(char32[])] Body
ALLOCATEMEMV D2 19 0; InlineFunc[void printf(char32[])] Params Start
WRITEPODSTRING D2 "[Test] Running %s\n"
PUSHR D2
LOAD D2 0.0002
PUSHVARGR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D2 0.0002; InlineFunc[int Game::GetGenericScript(char32[])] Params
GETGENERICSCRIPT D2; InlineFunc[int Game::GetGenericScript(char32[])] Body
STORE D2 0.0001
LOAD D2 0.0001; if() #10 Test
COMPAREV D2 -1
SETCMP D2 I==
COMPAREV D2 0
GOTOCMP 30 ==; Test 'if'
ALLOCATEMEMV D2 53 0; InlineFunc[void printf(char32[])] Params Start AND if() #10 Body Start
WRITEPODSTRING D2 "[Test] failed to start - could not find test script\n"
PUSHR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
GAMEEXIT; InlineFunc[void Game::Exit()] Body AND if() #10 Body End
LOAD D2 0.0001; InlineFunc[genericdata Game::LoadGenericData(int)] Params
LOADGENERICDATA D2; InlineFunc[genericdata Game::LoadGenericData(int)] Body Start
SETR D2 REFGENERICDATA; InlineFunc[genericdata Game::LoadGenericData(int)] Body End
STORE D2 0
SETV D2 1
PUSHR D2
LOAD D2 0
POP D3
PUSHR D2
PUSHR D3
POP D2
CASTBOOLI D2
POP REFGENERICDATA
SETR GENDATARUNNING D2
LOAD D2 0; while() #1 Test
PUSHR D2
POP REFGENERICDATA
SETR D2 GENDATARUNNING
COMPAREV D2 0
GOTOCMP 52 ==; Test 'while'
WAITFRAME; InlineFunc[void Waitframe()] Body AND while() #1 Body
GOTO 44; while() #1 End
CALLFUNC 0.0097; UsrFunc[void End()] Call AND [Opt:NeverRet]
REF_REMOVE 0.0002; [Opt:NeverRet] AND void run() Body End
PUSHR REFGENERICDATA; void run() Body Start
SETR D4 SP2
PUSHR D4
PUSHV 0; UsrFunc[void Init(int)] Params
CALLFUNC 0.0128; UsrFunc[void Init(int)] Call
POP D4
WAITFRAME; InlineFunc[void Waitframe()] Body
ALLOCATEMEMV D2 20 0; InlineFunc[void Trace(char32[])] Params Start
WRITEPODSTRING D2 "call_recursive: 100"; InlineFunc[void Trace(char32[])] Params End
TRACE6 D2; InlineFunc[void Trace(char32[])] Body Start
TRACE3; InlineFunc[void Trace(char32[])] Body End
PUSHR D4
PUSHV 100; UsrFunc[void call_recursive(int)] Params
CALLFUNC 0.0081; UsrFunc[void call_recursive(int)] Call
POP D4
ALLOCATEMEMV D2 22 0; InlineFunc[void Trace(char32[])] Params Start
WRITEPODSTRING D2 "call_recursive: 10000"; InlineFunc[void Trace(char32[])] Params End
TRACE6 D2; InlineFunc[void Trace(char32[])] Body Start
TRACE3; InlineFunc[void Trace(char32[])] Body End
PUSHR D4
PUSHV 10000; UsrFunc[void call_recursive(int)] Params
CALLFUNC 0.0081; UsrFunc[void call_recursive(int)] Call
POP D4
ALLOCATEMEMV D2 18 0; UsrFunc[void Fail(char32[])] Params Start
WRITEPODSTRING D2 "expected overflow"
PUSHR D2; UsrFunc[void Fail(char32[])] Params End
CALLFUNC 0.0104; UsrFunc[void Fail(char32[])] Call AND [Opt:NeverRet] AND [Opt:NeverRet] AND void run() Body End
SETR D4 SP2; Func[void call_recursive(int)] Body Start
LOAD D2 0; if() #1 Test
COMPAREV D2 0
SETCMP D2 I>
COMPAREV D2 0
GOTOCMP 93 ==; Test 'if'
PUSHR D4; if() #1 Body Start
LOAD D2 0; UsrFunc[void call_recursive(int)] Params Start
SUBV D2 1
PUSHR D2; UsrFunc[void call_recursive(int)] Params End
CALLFUNC 0.0081; UsrFunc[void call_recursive(int)] Call
POP D4; if() #1 Body End
LOAD D2 0; InlineFunc[void Trace(untyped)] Params
TRACER D2; InlineFunc[void Trace(untyped)] Body
POP D5
RETURNFUNC; Func[void call_recursive(int)] Body End
SETR D4 SP2; Func[void End()] Body Start
ALLOCATEMEMV D2 13 0; InlineFunc[void printf(char32[])] Params Start
WRITEPODSTRING D2 "[Test] done\n"
PUSHR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
GAMEEXIT; InlineFunc[void Game::Exit()] Body AND [Opt:NeverRet] AND Func[void End()] Body End
SETR D4 SP2; Func[void Fail(char32[])] Body Start
MARK_TYPE_STACK 0.0001 0
REF_INC 0
LOAD D2 0; InlineFunc[int strlen(char32[])] Params Start AND if() #1 Test
SETR D3 D2; InlineFunc[int strlen(char32[])] Params End
STRINGLENGTH D2 D3; InlineFunc[int strlen(char32[])] Body
COMPAREV D2 0
SETCMP D2 I>
COMPAREV D2 0
GOTOCMP 122 ==; Test 'if'
ALLOCATEMEMV D2 24 0; InlineFunc[void printf(char32[])] Params Start AND if() #1 Body Start
WRITEPODSTRING D2 "[Test] failed test: %s\n"
PUSHR D2
LOAD D2 0
PUSHVARGR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
GOTO 127; if() #1 Body End
ALLOCATEMEMV D2 20 0; InlineFunc[void printf(char32[])] Params Start AND if() #1 Else Start
WRITEPODSTRING D2 "[Test] failed test\n"
PUSHR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End AND if() #1 Else End
GAMEEXIT; InlineFunc[void Game::Exit()] Body AND [Opt:NeverRet] AND Func[void Fail(char32[])] Body End
SETR D4 SP2; Func[void Init(int)] Body Start
ALLOCATEMEMV D2 35 0; InlineFunc[void printf(char32[])] Params Start
WRITEPODSTRING D2 "[Test] started: dmap %d screen %d\n"
PUSHR D2
PUSHVARGV 5
LOAD D2 0
PUSHVARGR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End
LOAD D2 0; if() #9 Test
COMPAREV D2 0
SETCMP D2 I!=
COMPAREV D2 0
GOTOCMP 147 ==; Test 'if'
PUSHV 5; InlineFunc[void hero::Warp(int, int)] Params Start AND if() #9 Body Start
LOAD D2 0; InlineFunc[void hero::Warp(int, int)] Params End
POP D3; InlineFunc[void hero::Warp(int, int)] Body Start
WARPR D3 D2; InlineFunc[void hero::Warp(int, int)] Body End
WAITFRAME; InlineFunc[void Waitframe()] Body AND if() #9 Body End
POP D5
RETURNFUNC; Func[void Init(int)] Body End
